crypto/src/main/java/org/springframework/security/crypto/password/AbstractValidatingPasswordEncoder.java

```diff
@@ -32,21 +40,20 @@ public abstract class AbstractValidatingPasswordEncoder implements PasswordEncod
 
 	@Override
 	public final boolean matches(@Nullable CharSequence rawPassword, @Nullable String encodedPassword) {
-		if (rawPassword == null || rawPassword.length() == 0 || encodedPassword == null
-				|| encodedPassword.length() == 0) {
-			return false;
+		if (StringUtils.hasLength(rawPassword) && StringUtils.hasLength(encodedPassword)) {
+			return matchesNonNull(rawPassword.toString(), encodedPassword);
 		}
-		return matchesNonNull(rawPassword.toString(), encodedPassword);
+		return false;
 	}
 
 	protected abstract boolean matchesNonNull(String rawPassword, String encodedPassword);
 
 	@Override
 	public final boolean upgradeEncoding(@Nullable String encodedPassword) {
-		if (encodedPassword == null || encodedPassword.length() == 0) {
-			return false;
+		if (StringUtils.hasLength(encodedPassword)) {
```

Comment:
Please favor the simpler exits coming first:

```java
if (!StringUtils.hasLength(encodedPassword)) {
    return false;
}
return upgradeEncodingNonNull(encodedPassword);
```

This also has the added benefit of making the overall change smaller.